include/RAJA/exec-agency/forall_cuda_agency.hxx

```diff
@@ -0,0 +1,219 @@
+//
+// All rights reserved.
+//
+// This file is part of RAJA.
+//
+// For additional details, please also read RAJA/LICENSE.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the disclaimer below.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the disclaimer (as noted below) in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of the LLNS/LLNL nor the names of its contributors may
+//   be used to endorse or promote products derived from this software without
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/int_datatypes.hxx"
+#include "RAJA/fault_tolerance.hxx"
+#include "RAJA/segment_exec.hxx"
+#include "RAJA/internal/defines.hxx"
+
+#include "RAJA/exec-cuda/raja_cudaerrchk.hxx"
+#include "RAJA/exec-cuda/MemUtils_CUDA.hxx"
+
+namespace RAJA
+{
+//
+//////////////////////////////////////////////////////////////////////
+//
+// Agency CUDA kernel templates.
+//
+//////////////////////////////////////////////////////////////////////
+//
+template <size_t BLOCK_SIZE, bool Async, typename Iterable, typename LOOP_BODY, typename Agent, typename Worker>
+RAJA_INLINE void forall(agency_cuda_base<Agent, Worker, BLOCK_SIZE, Async>,
+                        Iterable&& iter,
+                        LOOP_BODY&& loop_body)
+{
+  beforeCudaKernelLaunch();
+  auto body = loop_body;
+  auto begin = std::begin(iter);
+  auto end = std::end(iter);
+  Index_type len = std::distance(begin, end);
+
+  size_t gridSize = RAJA_DIVIDE_CEILING_INT(len, BLOCK_SIZE);
+  gridSize = RAJA_MIN(gridSize, RAJA_CUDA_MAX_NUM_BLOCKS);
+
+  RAJA_FT_BEGIN;
+
+  if (len > 0) {
+    agency::bulk_invoke(Worker{}(gridSize, BLOCK_SIZE),
+                        [=] __device__ (Agent& self) {
```

Comment:
I was concerned that this might be subject to the same issue as in #7, however the tests pass and also use ListSegments (i.e. unified memory) so I'm not super worried.